tmpl/blog/mirage-entropy.md

```diff
@@ -0,0 +1,341 @@
+generates a bit more, hashes that, and uses the result as the next key. This is
+not to improve the statistical randomness, as this is already guaranteed by AES.
+Rather, it's a form of [forward secrecy][wiki-forward-secrecy]: an attacker who
+learns the secret key at some point would need to perform the [preimage
+attack][wiki-preimage-attack] on the hash function to figure out the earlier key
+and reconstruct the earlier outputs.
+
+[wiki-random-sequence]: https://en.wikipedia.org/wiki/Random_sequence
+[wiki-statistical-randomness]: https://en.wikipedia.org/wiki/Statistical_randomness
+[diehard-web]: http://www.stat.fsu.edu/pub/diehard/
+[marsaglia-pi-randomness]: http://interstat.statjournals.net/YEAR/2005/articles/0510005.pdf
+[wiki-algorithmic-randomness]: https://en.wikipedia.org/wiki/Algorithmically_random_sequence
+[wolfram-rule-30]: http://mathworld.wolfram.com/Rule30.html
+[mersenne-twister]: https://en.wikipedia.org/wiki/Mersenne_Twister
+[dgrb-boskovic]: http://qrbg.irb.hr/
+[fortuna]: https://www.schneier.com/fortuna.html
+[wiki-known-plaintext]: https://en.wikipedia.org/wiki/Known-plaintext_attack
+[wiki-forward-secrecy]: https://en.wikipedia.org/wiki/Forward_secrecy
+[wiki-preimage-attack]: https://en.wikipedia.org/wiki/Preimage_attack
+
+## Entropy
+
+Still, just like any other software RNG, Fortuna is just a deterministic
+algorithm. Its entire output is as unpredictable as its initial seed. From the
+information perspective, a RNG can only stretch what was unpredictable about its
+initial seed into an equally unpredictable sequence. The best generators do not
+give out more hints about their staring position, but they can never out-race
+the amount of unpredictability that was contained in the seed.
+
+We often call this unpredictable bit *entropy*, after the quality we are hoping
+it can bring into our RNG - the actual unpredictability. In a sense, by
+employing an algorithmic generator, we have just shifted the burden of being
+unpredictable around. But now we're cornered and have to search for it in the
+only place where a computer can find it -- in the physical world.
+
+A typical (kernel-level) RNG reaches out into the world around it through
+hardware interaction: as hardware events happen, various drivers tend to feed
+small packets of data, like time and hardware-specific state, into the RNG.
+These events are a product of the user interacts with the keyboard and mouse, of
+network packets arriving at the interface, or the hard drive asserting
+interrupts to signal the end of a DMA transfer. These events are combined
+together and used as a seed for a deterministic CSPRNG.
+
+In fact, describing them as a "seed" from which the entire sequence is unfolded
+is a deliberate oversimplification: what really happens is that the RNG is
+continuously fed with random events, which change its state as they arrive,
+while the requests for random bytes are served from the RNG. The RNG is used to
```

Comment:
I have difficulty in understanding all the "RNG" of this paragraph do point to the same notion (isn't one of these a PRNG) ?